pkg/controllers/oauthendpoints/oauth_endpoints_controller.go

```diff
@@ -0,0 +1,234 @@
+	}
+	return toHealthzURL(results), nil
+}
+
+func listOAuthServices(serviceLister corev1listers.ServiceLister, recorder events.Recorder) ([]string, error) {
+	var results []string
+	service, err := serviceLister.Services("openshift-authentication").Get("oauth-openshift")
+	if err != nil {
+		recorder.Warningf("OAuthServiceCheck", "failed to get oauth service: %v", err)
+		return nil, err
+	}
+	for _, port := range service.Spec.Ports {
+		results = append(results, net.JoinHostPort(service.Spec.ClusterIP, strconv.Itoa(int(port.Port))))
+	}
+	if len(results) == 0 {
+		return nil, fmt.Errorf("no valid oauth services found")
+	}
+	return toHealthzURL(results), nil
+}
+
+func listOAuthRoutes(routeLister routev1listers.RouteLister, recorder events.Recorder) ([]string, error) {
+	var results []string
+	route, err := routeLister.Routes("openshift-authentication").Get("oauth-openshift")
+	if err != nil {
+		recorder.Warningf("OAuthRouteCheck", "failed to get oauth route: %v", err)
+		return nil, err
+	}
+	for _, ingress := range route.Status.Ingress {
+		if len(ingress.Host) > 0 {
+			results = append(results, ingress.Host)
+		}
+	}
+	if len(results) == 0 {
+		recorder.Warningf("OAuthRouteCheck", "route status does not have host address")
+		return nil, fmt.Errorf("route status does not have host address")
+	}
+	return toHealthzURL(results), nil
+}
+
+func getOAuthRouteTLSConfig(secretLister corev1listers.SecretLister, ingressLister configv1lister.IngressLister, systemCABundle []byte, recorder events.Recorder) (*tls.Config, error) {
+	ingress, err := ingressLister.Get("cluster")
+	if err != nil {
+		recorder.Warningf("OAuthRouteSecret", "failed to get ingress config: %v", err)
+		return nil, err
+	}
+	if len(ingress.Spec.Domain) == 0 {
+		return nil, fmt.Errorf("ingress config domain cannot be empty")
+	}
+
+	routerSecret, err := secretLister.Secrets("openshift-authentication").Get("v4-0-config-system-router-certs")
+	if err != nil {
+		recorder.Warningf("OAuthRouteSecret", "failed to get oauth route ca cert: %v", err)
+		return nil, err
+	}
+
+	// find the domain that matches our route
+	if _, ok := routerSecret.Data[ingress.Spec.Domain]; !ok {
+		klog.Infof("unable to find router certs for domain %s", ingress.Spec.Domain)
+		return nil, nil
+	}
+
+	rootCAs := x509.NewCertPool()
+	if ok := rootCAs.AppendCertsFromPEM(routerSecret.Data[ingress.Spec.Domain]); !ok {
+		klog.Infof("failed to parse router certs for domain %s", ingress.Spec.Domain)
+		return nil, nil
+	}
+
+	if len(systemCABundle) > 0 {
+		if ok := rootCAs.AppendCertsFromPEM(systemCABundle); !ok {
+			klog.Infof("failed to parse system ca bundle")
+			return nil, nil
+		}
+	}
+
+	return &tls.Config{
+		RootCAs: rootCAs,
+	}, nil
+}
+
+func getOAuthEndpointTLSConfig(secretLister corev1listers.SecretLister, recorder events.Recorder) (*tls.Config, error) {
+	serviceSecret, err := secretLister.Secrets("openshift-authentication").Get("v4-0-config-system-serving-cert")
```

Comment:
shouldn't we load a public cert of the authority that signed the serving cert for `oauth-server` instead?